ExchangeSharp/API/Exchanges/Digifnex/ExchangeDigifinexAPI.cs

```diff
@@ -0,0 +1,514 @@
+﻿using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace ExchangeSharp
+{
+
+    public partial class ExchangeDigifinexAPI : ExchangeAPI
+    {
+        public override string BaseUrl { get; set; } = "https://openapi.digifinex.vip/v3";
+        public override string BaseUrlWebSocket { get; set; } = "wss://openapi.digifinex.com/ws/v1/";
+        int websocketMessageId = 0;
+        string timeWindow;
+
+        public ExchangeDigifinexAPI()
+        {
+            MarketSymbolSeparator = "_";
+            MarketSymbolIsReversed = false;
+            MarketSymbolIsUppercase = false;
+            WebSocketOrderBookType = WebSocketOrderBookType.FullBookFirstThenDeltas;
+            NonceStyle = NonceStyle.UnixSeconds;
+        }
+
+        #region ProcessRequest 
+
+        protected override async Task OnGetNonceOffset()
+        {
+            try
+            {
+                var start = CryptoUtility.UtcNow;
+                JToken token = await MakeJsonRequestAsync<JToken>("/time");
+                DateTime serverDate = DateTimeOffset.FromUnixTimeSeconds((long)token["server_time"]).UtcDateTime;
+                var end = CryptoUtility.UtcNow;
+                var now = start + TimeSpan.FromMilliseconds((end - start).TotalMilliseconds / 2);
+                var timeFaster = now - serverDate;
+                if (timeFaster <= TimeSpan.Zero)
+                {
+                    timeWindow = (timeFaster.Negate().TotalSeconds*10).ToString();
+                    NonceOffset = TimeSpan.FromSeconds(2.5);
+                }
+                else
+                    NonceOffset = now - serverDate; // how much time to substract from Nonce when making a request
+            }
+            catch (Exception)
+            { }
+        }
+
```

Comment:
I've added a checkin just now that implements a OnGetNonceOffset method and adds 3 new properties for the exchange to implement if the api supports a server timestamp:

NonceEndPoint, NonceEndPointField, NonceEndPointStyle

Can you explain why you have logic to divide by 2, multiply by 10 and then use a 2.5 second offset? What is timeWindow? I'd like to see if we can move some of this into the base OnGetNonceOffset method to avoid future exchanges needing this kind of logic.

Reply:
/ 2 is to take the average of send and recv transfer overhead and estimate the time when the server take its time.

The timeWindow is a parameter to tell the Digifinex server how much time span it can tolerate that the client time is behind of its server time. However, the server cannot tolerate that client time is ahead of server time. When NonceOffset is negative, I reset it to +2.5 seconds to prevent ExchangeSharp to send nonce that ahead of the server time. And I use the timeWindow to let the server tolerate the client time behind. x10 is just a number that is kind of *safe* (see the [doc](https://docs.digifinex.vip/en-ww/v3/) and search for `ACCESS-RECV-WINDOW`).

When the client time is ahead of the server time (NonceOffset > 0), I just follow the usual process flow of ExchangeSharp.

Reply:
Thanks for the clarification. We'll have to see in the future if we can move this into the base class, but for now this method looks good.